orc/src/main/java/org/apache/iceberg/orc/ExpressionToSearchArgument.java

```diff
@@ -0,0 +1,355 @@
+    }
+  }
+
+  @FunctionalInterface
+  interface Action {
+    void invoke();
+  }
+
+  private PredicateLeaf.Type type(Type icebergType) {
+    switch (icebergType.typeId()) {
+      case BOOLEAN:
+        return PredicateLeaf.Type.BOOLEAN;
+      case INTEGER:
+      case LONG:
+      case TIME:
+        return PredicateLeaf.Type.LONG;
+      case FLOAT:
+      case DOUBLE:
+        return PredicateLeaf.Type.FLOAT;
+      case DATE:
+        return PredicateLeaf.Type.DATE;
+      case TIMESTAMP:
+        return PredicateLeaf.Type.TIMESTAMP;
+      case STRING:
+        return PredicateLeaf.Type.STRING;
+      case DECIMAL:
+        return PredicateLeaf.Type.DECIMAL;
+      default:
+        throw new UnsupportedOperationException("Type " + icebergType + " not supported in ORC SearchArguments");
+    }
+  }
+
+  private <T> Object literal(Type icebergType, T icebergLiteral) {
+    switch (icebergType.typeId()) {
+      case BOOLEAN:
+      case LONG:
+      case TIME:
+      case DOUBLE:
+        return icebergLiteral;
+      case INTEGER:
+        return ((Integer) icebergLiteral).longValue();
+      case FLOAT:
+        return ((Float) icebergLiteral).doubleValue();
+      case STRING:
+        return icebergLiteral.toString();
+      case DATE:
+        return Date.valueOf(LocalDate.ofEpochDay((Integer) icebergLiteral));
+      case TIMESTAMP:
+        long microsFromEpoch = (Long) icebergLiteral;
+        return Timestamp.from(Instant.ofEpochSecond(Math.floorDiv(microsFromEpoch, 1_000_000),
+            (microsFromEpoch % 1_000_000) * 1_000));
+      case DECIMAL:
+        return new HiveDecimalWritable(HiveDecimal.create((BigDecimal) icebergLiteral, false));
+      default:
+        throw new UnsupportedOperationException("Type " + icebergType + " not supported in ORC SearchArguments");
+    }
+  }
+
+  /**
+   * Generates mapping from field IDs to fully qualified column names compatible with ORC convention for a given
+   * {@link Schema}
+   *
+   * This visitor also enclose column names in backticks i.e. ` so that ORC can correctly parse column names with
+   * special characters. A comparison of ORC convention with Iceberg convention is provided below
+   * <pre>
+   *                                      Iceberg           ORC
+   * field                                field             field
+   * struct -> field                      struct.field      struct.field
+   * list -> element                      list.element      list._elem
+   * list -> struct element -> field      list.field        list._elem.field
+   * map -> key                           map.key           map._key
+   * map -> value                         map.value         map._value
+   * map -> struct key -> field           map.key.field     map._key.field
+   * map -> struct value -> field         map.field         map._value.field
+   * </pre>
+   */
+  static class IdToQuotedColumnName extends TypeUtil.CustomOrderSchemaVisitor<Map<Integer, String>> {
```

Comment:
As a follow-up, you might consider rewriting this to use the new before/after methods instead of the custom order visitor. We just converted `IndexByName` and the implementation is much easier to read and maintain.

Reply:
Done